README.md
# Assetylene, a plugin for Movable Type

This plugin provides a new "Caption" field when inserting an asset into a post,
and the ability to customize the HTML markup produced for publishing MT assets.

# Prerequisites

* Movable Type 4.x or 5.x

# Installation

To install this plugin follow the instructions found here:

http://tinyurl.com/easy-plugin-install

# Usage

Create a Template Module named "Asset Insertion", and that template will then
be used to create the HTML for any asset you embed into a post.

There are a number of template variables available to this template:

* `upload_html`: This is the original markup MT constructed for the inserted
  asset.
* `enclose`: 1 or 0, depending on whether the asset is to be wrapped with a
  'form' tag.
* `include`: 1 or 0, depending on whether the image is to be displayed or simply
  linked to.
* `thumb`: 1 or 0, depending on whether the inline image is a thumbnail of the
  original.
* `align`: The value of the alignment option (`none`, `left`, `right`, or
  `center`).
* `popup`: 1 or 0, depending on whether the thumbnail image is linked to a popup
  of the full image.
* `caption`: The value entered for the caption on the insert options dialog.
* `label`: The "label" value of the asset being inserted.
* `description`: The "description" value of the asset being inserted.
* `asset_id`: The numeric ID of the asset being inserted.
* `form_tag`: The opening HTML form tag MT creates to enclose the asset. If you
  output this, be sure to add a closing form tag as well. (Note that the form
  tag, classes, and styling were only used in MT4.2x and earlier.)
* `form_class`: The CSS class name(s) applied to the form tag.  (Note that the
  form tag, classes, and styling were only used in MT4.2x and earlier.)
* `form_style`: The CSS styling applied to the form tag. (Note that the form
  tag, classes, and styling were only used in MT4.2x and earlier.)
* `img_tag`: The MT-constructed img tag being inserted (this is unset if no
  image is actually inserted and linked to instead).
* `img_alt`: The "alt" attribute of the inserted image.
* `img_height`: The numeric height of the inserted image.
* `img_width`: The numeric width of the inserted image.
* `img_src`: The url of the inserted image.
* `img_style`: Any CSS styling applied to the inline image.
* `img_class`: Any CSS class name(s) applied to the inline image.
* `a_tag`: The "a" tag produced to link the image (if any).
* `a_href`: The "href" attribute value of the "a" tag.
* `a_onclick`: The "onclick" attribute value of the "a" tag.

Also, the asset itself is in context, so any of the MT Asset tags may be used
to generate the asset markup. The current blog is also in context.

An example template can be found in:

    plugins/Assestylene/asset_insertion.mtml

You can use this as a basis for your own custom template.

# License

This plugin has been released under the terms of the GNU Public License,
version 2.0.


CHANGES.txt
# Version 1.12 - Jan 04, 2011 (naoaki011)

    * Add size limit variables for image & thumbnail

# Version 1.11 - Dec 11, 2010 (naoaki011)

    * Support multi output results select in asset insertion

# Version 1.10 - Sep 22, 2010 (naoaki011)

    * Add support for Lightbox effect Handling

# Version 1.02 - May 24, 2009 (byrnereese)

    * Improved some regexes so they are less greedy
      http://code.sixapart.com/trac/mtplugins/changeset/1534/trunk/Assetylene

# Version 1.01 - May 1, 2009 (byrnereese)

    * Making small change to the size of the caption box so that it does not
      disrupt the asset options dialog so much.
      http://code.sixapart.com/trac/mtplugins/changeset/1524/trunk/Assetylene

    * Fixed bug in which many of the variables (img_src, etc) were not
      properly being set, and added an onfocus javascript hook for the
      caption textarea.
      http://code.sixapart.com/trac/mtplugins/changeset/1525/trunk/Assetylene
  
# Version 1.0 - December 8, 2008 (bradchoate)

    * Initial release; support for MT 4.2 and later.


plugins/Assetylene/lib/Assetylene/CMS.pm
# This program is distributed under the terms of the
# GNU General Public License, version 2.
#
# $Id: CMS.pm 1534 2009-05-24 23:52:58Z breese $

package Assetylene::CMS;

use strict;
use MT 4;
use Assetylene::L10N;
use MT::Util qw( encode_html );

sub asset_options_image {
    my ($cb, $app, $param, $tmpl) = @_;

    # Assertions:
    # 'asset_id' template parameter must be present.
    my $asset_id = $param->{asset_id} or return;

    # Asset object must be loadable
    my $asset = MT::Asset->load( $asset_id ) or return;

    # The 'image_alignment' MT template node must be in
    # the template we're working with to add our field above it.
    my $el = $tmpl->getElementById('image_alignment')
        or return;

# Caption >
    my $opt = $tmpl->createElement('app:setting', {
        id => 'image_caption',
        label => MT->translate('Caption'),
        label_class => 'no-header',
        hint => '',
        show_hint => 0,
    });

    require MT::Util;
    # Encode any special characters as HTML entities, since this
    # description is being placed in an HTML textarea:
    my $caption_safe = MT::Util::encode_html( $asset->description );

    # Contents of the app:setting tag:
    $opt->innerHTML(<<HTML);
    <input type="checkbox" id="insert_caption" name="insert_caption"
        value="1" />
    <label for="insert_caption"><__trans_section component="Assetylene"><__trans phrase='Insert a caption?'></__trans_section></label>
    <div class="textarea-wrapper"><textarea name="caption" style="height: 36px;" rows="2" cols=""
        onfocus="getByID('insert_caption').checked=true; return false;"
        class="full-width">$caption_safe</textarea></div>
HTML
    # Insert new field above the 'image_alignment' field:
    $tmpl->insertBefore($opt, $el);

#< Caption
# Pattern >

    $opt = $tmpl->createElement('app:setting', {
        id => 'asset_optins',
        label => MT->translate('Asset Options'),
        label_class => 'no-header',
        hint => '',
        show_hint => 0,
    });

    my $blog_id = $app->param('blog_id');
    my $plugin = MT->component("Assetylene");
    my $scope = "blog:".$blog_id;
    my $insert_options = '';

    my $pattern_name1 = MT::Util::encode_html($plugin->get_config_value('pattern1',$scope),1);
    if ($pattern_name1) {
        $insert_options .= '<option value="1">' . $pattern_name1 . '</option>' . "\n";
    }
    my $pattern_name2 = MT::Util::encode_html($plugin->get_config_value('pattern2',$scope),1);
    if ($pattern_name2) {
        $insert_options .= '<option value="2">' . $pattern_name2 . '</option>' . "\n";
    }
    my $pattern_name3 = MT::Util::encode_html($plugin->get_config_value('pattern3',$scope),1);
    if ($pattern_name3) {
        $insert_options .= '<option value="3">' . $pattern_name3 . '</option>' . "\n";
    }
    my $pattern_name4 = MT::Util::encode_html($plugin->get_config_value('pattern4',$scope),1);
    if ($pattern_name4) {
        $insert_options .= '<option value="4">' . $pattern_name4 . '</option>' . "\n";
    }
    my $pattern_name5 = MT::Util::encode_html($plugin->get_config_value('pattern5',$scope),1);
    if ($pattern_name5) {
        $insert_options .= '<option value="5">' . $pattern_name5 . '</option>' . "\n";
    }

    if ($insert_options) {
        $opt->innerHTML(<<HTML);
    <label for="pattern"><__trans_section component="Assetylene"><__trans phrase='Insertion Pattern'></__trans_section></label>
    <select name="pattern">
        $insert_options
    </select><br />
HTML
        $tmpl->insertBefore($opt, $el);
    }

#< Pattern
# limit_size >

    my $limit_thumbnail_width = $plugin->get_config_value('limit_thumbnail_width',$scope);
    my $max_thumbnail_width = $plugin->get_config_value('max_thumbnail_width',$scope);
    my $limit_thumbnail_height = $plugin->get_config_value('limit_thumbnail_height',$scope);
    my $max_thumbnail_height = $plugin->get_config_value('max_thumbnail_height',$scope);

    my $limit_width = $plugin->get_config_value('limit_width',$scope);
    my $max_width = $plugin->get_config_value('max_width',$scope);
    my $limit_height = $plugin->get_config_value('limit_height',$scope);
    my $max_height = $plugin->get_config_value('max_height',$scope);

#    my $myelm = $tmpl->getElementById('create_thumbnail-field')
#        or return;


# < limit_size
# lightbox >

    $opt = $tmpl->createElement('app:setting', {
        id => 'asset_lightbox',
        label => MT->translate('Lightbox'),
        label_class => 'no-header',
        hint => '',
        show_hint => 0,
    });

    use MT::Blog;
    my $blog = MT::Blog->load($blog_id) or die;
    my $themeid = $blog->theme_id;
    if ($themeid ne 'mtVicunaSimple') {

        my $insert_options = '';
        my $lb_select1 = $plugin->get_config_value('lb_select1',$scope);
        my $lightbox_selector1 = MT::Util::encode_html($plugin->get_config_value('lightbox_selector1',$scope),1);
        if (($lb_select1) && ($lightbox_selector1)) {
            $insert_options .= '<option value="' . $lightbox_selector1 . '">' . $lightbox_selector1 . '</option>' . "\n";
        }
        my $lb_select2 = $plugin->get_config_value('lb_select2',$scope);
        my $lightbox_selector2 = MT::Util::encode_html($plugin->get_config_value('lightbox_selector2',$scope),1);
        if (($lb_select2) && ($lightbox_selector2)) {
            $insert_options .= '<option value="' . $lightbox_selector2 . '">' . $lightbox_selector2 . '</option>' . "\n";
        }
        my $lb_select3 = $plugin->get_config_value('lb_select3',$scope);
        my $lightbox_selector3 = MT::Util::encode_html($plugin->get_config_value('lightbox_selector3',$scope),1);
        if (($lb_select3) && ($lightbox_selector3)) {
            $insert_options .= '<option value="' . $lightbox_selector3 . '">' . $lightbox_selector3 . '</option>' . "\n";
        }
        my $lb_select4 = $plugin->get_config_value('lb_select4',$scope);
        my $lightbox_selector4 = MT::Util::encode_html($plugin->get_config_value('lightbox_selector4',$scope),1);
        if (($lb_select4) && ($lightbox_selector4)) {
            $insert_options .= '<option value="' . $lightbox_selector4 . '">' . $lightbox_selector4 . '</option>' . "\n";
        }
        if ($insert_options eq '') {
            $insert_options .= '<option value="rel=&quot;lightbox&quot;">rel=&quot;lightbox&quot;</option>' . "\n";
        }

        $opt->innerHTML(<<HTML);
        <div>
            <input type="checkbox" id="insert_lightbox" name="insert_lightbox"
                onclick="if(this.checked){document.getElementById('create_thumbnail').checked=true;
                  document.getElementById('thumb_width').focus();
                  document.getElementById('thumb_width').value=$max_thumbnail_width;
                }else{
                  document.getElementById('create_thumbnail').checked=false;
                }"
                value="1"<mt:if name="make_thumb"> checked="checked" </mt:if> />
            <label for="insert_lightbox"><__trans_section component="Assetylene"><__trans phrase='Use Lightbox Effect'></__trans_section></label>
            <select id="insert_class" name="insert_class">
                $insert_options
            </select>
        </div>
HTML
    }
    $tmpl->insertBefore($opt, $el);
# < lightbox

    # Force the tokens of the template to be reprocessed now that
    # we've manipulated it:
    $tmpl->rescan();
}

sub asset_insert {
    my ($cb, $app, $param, $tmpl) = @_;

    my $blog_id = $app->param('blog_id');
    my $upload_html = $param->{ upload_html };

    use MT::Blog;
    my $blog = MT::Blog->load($blog_id) or die;
    my $themeid = $blog->theme_id;
    my $plugin = MT->component("Assetylene");
    my $scope = "blog:".$blog_id;
    if ($themeid ne 'mtVicunaSimple') {
        my $cleanup_insert = $plugin->get_config_value('cleanup_insert',$scope);
        if ($cleanup_insert) {
            my $rightalign_class = $plugin->get_config_value('rightalign_class',$scope);
            my $centeralign_class = $plugin->get_config_value('centeralign_class',$scope);
            my $leftalign_class = $plugin->get_config_value('leftalign_class',$scope);
            my $wrap;
            if ($cleanup_insert == '1' || $cleanup_insert == '') {
                if ($upload_html =~ / class=\"mt-image-left\"/) {
                    $wrap = '<p class="'.$leftalign_class.'">';
                    if ($cleanup_insert == '') {
                        $wrap = '<p class="img_L">';
                    }
                }
                if ($upload_html =~ / class=\"mt-image-right\"/) {
                    $wrap = '<p class="'.$rightalign_class.'">';
                    if ($cleanup_insert == '') {
                        $wrap = '<p class="img_R">';
                    }
                }
                if ($upload_html =~ / class=\"mt-image-center\"/) {
                    if ($centeralign_class) {
                        $wrap = '<p class="'.$centeralign_class.'">';
                    } else {
                        $wrap = '<p>';
                    }
                    if ($cleanup_insert == '') {
                        $wrap = '<p>';
                    }
                }
                $upload_html =~ s/ class=\"mt-image-(none|right|left|center)\"//g;
            }
            if ($cleanup_insert == '2') {
                $upload_html =~ s/ class=\"mt-image-none\"//i;
                $rightalign_class = ' class="'.$rightalign_class.'"';
                $upload_html =~ s/ class=\"mt-image-right\"/$rightalign_class/g;
                $leftalign_class = ' class="'.$leftalign_class.'"';
                $upload_html =~ s/ class=\"mt-image-left\"/$leftalign_class/g;
                if ($centeralign_class) {
                    $centeralign_class = ' class="'.$centeralign_class.'"';
                }
                $upload_html =~ s/ class=\"mt-image-center\"/$centeralign_class/g;
            } else {
                $upload_html =~ s/ class=\"mt-image-(none|right|left|center)\"//g;
            }
            $upload_html =~ s/ style=\"\"//i;
            $upload_html =~ s/ style=\"float\: (right|left)\; margin\: 0 (0|20px) 20px (0|20px)\;\"//i;
            $upload_html =~ s/ style=\"text-align\: center\; display\: block\; margin\: 0 auto 20px\;\"//i;
            if ($wrap) {
                $upload_html = $wrap.$upload_html.'</p>';
            }
        }
        my $insert_class = $app->param('insert_class');
        if ( $app->param('insert_lightbox') ) {
            $insert_class = '<a '.$insert_class;
            $upload_html =~ s/<a/$insert_class/g;
        }
    }

    my $blog = $app->blog;

    # Assertions:
    # Load the user-defined "Asset Insertion" template module.
    # Currently, this template must be named in English. Look both
    # at the blog and system level for this template.
    my $insert_tmpl = $app->model('template')->load({
        name => 'Asset Insertion', type => 'custom',
        blog_id => [ $blog->id, 0 ] });
    return unless $insert_tmpl;

    my $asset = $tmpl->context->stash('asset');

    # Collect all the elements of the MT generated asset markup
    # so they can be manipulated indepdendently by the user-defined
    # template:
    my $html = $upload_html;
    my ($img_tag) = $html =~ /(<img\b[^>]+?>)/s;
    my ($a_tag) = $html =~ /(<a\b[^>]+?>)/s;
    my ($form_tag) = $html =~ /(<form[^>]+?>)/s;

    $param->{enclose} = 1 if $form_tag;
    $param->{include} = 1 if $app->param('include');
    $param->{thumb} = 1 if $app->param('thumb');
    ($param->{align}) = $app->param('align') =~ m/(\w+)/;
    $param->{caption} = $app->param('insert_caption') ? $app->param('caption') : '';
    $param->{popup} = 1 if $app->param('popup');

    $param->{label} = $asset->label;
    $param->{description} = $asset->description;
    $param->{asset_id} = $asset->id;

    $param->{a_tag} = $a_tag;
    ($param->{a_href}) = $a_tag =~ /\bhref="(.+?)"/s;
    ($param->{a_onclick}) = $a_tag =~ /\bonclick="(.+?)"/s;

    ($param->{a_style}) = $a_tag =~ /\bstyle="([^\"]+)"/s;
    ($param->{a_class}) = $a_tag =~ /\bclass="([^\"]+)"/s;
    ($param->{a_title}) = $a_tag =~ /\btitle="([^\"]+)"/s;
    ($param->{a_rel}) = $a_tag =~ /\brel="([^\"]+)"/s;

    $param->{form_tag} = $form_tag;
    ($param->{form_style}) = $form_tag =~ /\bstyle="([^\"]+)"/s;
    ($param->{form_class}) = $form_tag =~ /\bclass="([^\"]+)"/s;

    $param->{img_tag} = $img_tag;
    ($param->{img_height}) = $img_tag =~ /\bheight="(\d+)"/;
    ($param->{img_width}) = $img_tag =~ /\bwidth="(\d+)"/;
    ($param->{img_src}) = $img_tag =~ /\bsrc="([^\"]+)"/s;
    ($param->{img_style}) = $img_tag =~ /\bstyle="([^\"]+)"/s;
    ($param->{img_class}) = $img_tag =~ /\bclass="([^\"]+)"/s;
    ($param->{img_alt}) = $img_tag =~ /\balt="([^\"]+)"/s;

    $param->{pattern} = $app->param('pattern');

    my $limit_thumbnail_width = $plugin->get_config_value('limit_thumbnail_width',$scope);
    my $max_thumbnail_width = $plugin->get_config_value('max_thumbnail_width',$scope);
    my $limit_thumbnail_height = $plugin->get_config_value('limit_thumbnail_height',$scope);
    my $max_thumbnail_height = $plugin->get_config_value('max_thumbnail_height',$scope);

    $param->{limit_thumbnail_width} = 1 if $limit_thumbnail_width;
    $param->{max_thumbnail_width} = $max_thumbnail_width;
    $param->{limit_thumbnail_height} = 1 if $limit_thumbnail_height;
    $param->{max_thumbnail_height} = $max_thumbnail_height;

    my $limit_width = $plugin->get_config_value('limit_width',$scope);
    my $max_width = $plugin->get_config_value('max_width',$scope);
    my $limit_height = $plugin->get_config_value('limit_height',$scope);
    my $max_height = $plugin->get_config_value('max_height',$scope);

    $param->{limit_width} = 1 if $limit_width;
    $param->{max_width} = $max_width;
    $param->{limit_height} = 1 if $limit_height;
    $param->{max_height} = $max_height;

    $insert_tmpl->param( $param );

    my $ctx = $insert_tmpl->context;
    $ctx->stash('blog', $blog);
    $ctx->stash('blog_id', $blog->id);
    $ctx->stash('local_blog_id', $blog->id);
    $ctx->stash('asset', $asset);

    # Process the user-defined template:
    my $new_html = $insert_tmpl->output;

    my $remove_blank = $plugin->get_config_value('remove_blank',$scope);
    if ($remove_blank) {
        $new_html =~ s/\s*\n+/\n/g;
    }

    if (defined($new_html)) {
        # Replace the MT generated asset markup with the user-defined
        # markup:
        $param->{upload_html} = $new_html;
    }
    else {
        # Template build error: die, so this gets logged (we're in a
        # callback, so it won't be surfaced to the user unfortunately)
        die "Error from Asset Insertion module: " . $insert_tmpl->errstr;
    }
}

sub doLog {
    my ($msg) = @_; 
    return unless defined($msg);
    require MT::Log;
    my $log = MT::Log->new;
    $log->message($msg) ;
    $log->save or die $log->errstr;
}

1;
